source/engines/rollEngine.js

```diff
@@ -1,548 +1,771 @@
+	#dieSides;
+
+	constructor(dieSides) {
+		super(Math.floor(Math.random() * dieSides) + 1);
+		this.#dieSides = dieSides;
+	}
+
+	getMaxValue() {
+		return this.#dieSides;
+	}
+
+	toString(frac = false) {
+		return frac ? `${this.getResult()}/${this.getMaxValue()}` : `${this.getResult()}`;
+	}
+}
+
+class StringResult extends Result {
+	#optList;
+
+	constructor(result, optList) {
+		super(result);
+		this.#optList = optList;
+	}
+
+	getMaxValue() {
+		return this.getResult();
+	}
+
+	toString(frac = false) {
+		return frac ? `${this.getResult()} from the list ${this.#optList}` : this.getResult();
+	}
+}
+
+const two_ops = { // Mathematical operations we currently support for rolls that use two arguments
+	ADD: '+',
+	SUBTRACT: '-',
+	MULTIPLY: '*'
+}
+
+const keep_ops = { // Operations for keeping or dropping rolls of a result set
+	KEEP: 'k',
+	KEEP_HIGHEST: 'kh',
+	DROP_LOWEST: 'dl',
+	DROP: 'd',
+	DROP_HIGHEST: 'dh',
+	KEEP_LOWEST: 'kl'
+}
+
+class ResultSet {
+
+	constructor() {}
+
+	static #surroundedByParens(str, symbolPos) {
+		var rightParensUnenclosed = 0,
+			leftParensUnenclosed = 0;
+		var leftSide = str.slice(0,symbolPos),
+			rightSide = str.slice(symbolPos+1,str.length-1);
+		for (var i = leftSide.length - 1; i >= 0 ; i-=1) {
+			if (leftSide[i] == ')' && leftParensUnenclosed <= 0) {
+				leftParensUnenclosed -= 1;
+			} else if (leftSide[i] == '(') {
+				leftParensUnenclosed += 1;
+			}
+		}
+		for (var i = 0; i < rightSide.length; i+=1) {
+			if (rightSide[i] == '(' && rightParensUnenclosed <= 0) {
+				rightParensUnenclosed -= 1;
+			} else if (rightSide[i] == ')') {
+				rightParensUnenclosed += 1;
+			}
+		}
+		return rightParensUnenclosed > 0 && leftParensUnenclosed > 0;
+	}
+
+	static #parseStringList(str) {
+		let strList = [];
+		let newEntry = "";
+		let startQuote = false;
+		let endQuote = false;
+		let spaceCount = 0;
+		for (var i = 0; i < str.length; i++) {
+			if (
+				(str[i] == '[' && i == 0) ||
+				(!(startQuote && endQuote) && str[i] == ']' && i == str.length-1) ||
+				(startQuote && endQuote && str[i] == ' ')
+				) {
+					if (str[i] == ' ') {
+						spaceCount++;
+					}
+				continue;
+			} else if (startQuote && endQuote && str[i] == ']' && i == str.length-1) {
+				strList.push(newEntry.trim());
+				newEntry = "";
+				startQuote = false;
+				endQuote = false;
+				spaceCount = 0;
+				continue;
+			} else if (startQuote && endQuote && str[i] === ',') {
+				strList.push(newEntry.trim());
+				newEntry = "";
+				startQuote = false;
+				endQuote = false;
+				spaceCount = 0;
+				continue;
+			} else if (startQuote && endQuote) {
+				newEntry = `"${newEntry}"${' '.repeat(spaceCount)}${str[i]}`;
+				startQuote = false;
+				endQuote = false;
+				spaceCount = 0;
+				continue;
+			} else if (startQuote && !endQuote && str[i] == "\"") {
+				endQuote = true;
+				continue;
+			} else if (!startQuote && !endQuote && str[i] == "\"") {
+				startQuote = true;
+				continue;

```

Comment:
```suggestion
			} else if (str[i] == "\"" && !endQuote) {
				if (startQuote) {
					endQuote = true;
					continue;
				} else {
					startQuote = true;
					continue;
				}
```
Slightly less evaluation on average (and reads clearer IMO).

Reply:
I disagree with this being clearer. The point of this structure is i'm trying to create distinct cases (like a switch/case). and having additional logic beyond the `case` makes it much more complicated to conceptually separate distinct cases. (There is an argument that these aren't exactly distinct cases, but the goal here was to force them to be.)

Reply:
Funnily enough, switching to trinary nukes this whole thing.

Reply:
I'm not feeling like implementing a new javascript datatype today, but maybe worth the backlog.